File: web/src/lib/bible.ts
/**
 * Client for the deployed Amharic Bible content API.
 *
 * The API has no CORS headers, so browser calls in the web app go through the
 * Next.js rewrite at /bible-api/* (see next.config.ts) which proxies to Koyeb.
 * Server-side calls can hit the API directly.
 */

const DIRECT =
  process.env.NEXT_PUBLIC_BIBLE_API ??
  "https://faithful-marni-anatoli-b7663357.koyeb.app";

// In the browser we use the same-origin proxy to dodge CORS; on the server, direct.
const BASE =
  typeof window === "undefined" ? DIRECT : "/bible-api";

export type BookRef = { num: number; name: string };

export type Verse = string;

export type Chapter = {
  chapter: string;
  title: string;
  verses: Verse[];
};

export type Book = {
  title: string;
  abbv: string;
  chapters: Chapter[];
};

async function get<T>(path: string): Promise<T> {
  const isServer = typeof window === "undefined";
  let lastErr: unknown;
  // Retry with a short per-attempt timeout — the upstream host occasionally
  // stalls on IPv6/Cloudflare, so fail fast and try again rather than hang.
  const MAX = 5;
  for (let attempt = 0; attempt < MAX; attempt++) {
    try {
      const ctrl = new AbortController();
      const timer = setTimeout(() => ctrl.abort(), 7000);
      const res = await fetch(`${BASE}${path}`, {
        signal: ctrl.signal,
        // Server caches immutable content; client relies on the in-memory Map
        // and skips the HTTP cache so a once-bad/partial response can't stick.
        ...(isServer
          ? { next: { revalidate: 60 * 60 * 24 } }
          : { cache: "no-store" as RequestCache }),
      });
      clearTimeout(timer);
      if (!res.ok) throw new Error(`Bible API ${res.status} for ${path}`);
      return (await res.json()) as T;
    } catch (e) {
      lastErr = e;
      if (attempt < MAX - 1)
        await new Promise((r) => setTimeout(r, 300 * (attempt + 1)));
    }
  }
  throw lastErr instanceof Error ? lastErr : new Error(`Bible API failed for ${path}`);
}

/** All 66 books as { num, name }, parsed from the API's dictionary array. */
export async function getBooks(): Promise<BookRef[]> {
  const arr = await get<string[]>("/book");
  return arr
    .slice(1) // index 0 is a header string
    .map((s) => {
      const m = s.match(/^(\d+)\s*:\s*(.+)$/);
      return m ? { num: parseInt(m[1], 10), name: m[2].trim() } : null;
    })
    .filter((b): b is BookRef => b !== null)
    .sort((a, b) => a.num - b.num);
}

/** A whole book (all chapters + verses). */
export async function getBook(num: number): Promise<Book> {
  return get<Book>(`/book/${num}`);
}

/** A single chapter. */
export async function getChapter(book: number, chapter: number): Promise<Chapter> {
  return get<Chapter>(`/book/${book}/chapter/${chapter}`);
}

/** Static metadata: book name lookups without a network call after first load. */
export function formatRef(
  bookName: string,
  chapter: number,
  verseStart: number,
  verseEnd?: number,
): string {
  const range =
    verseEnd && verseEnd !== verseStart
      ? `${verseStart}-${verseEnd}`
      : `${verseStart}`;
  return `${bookName} ${chapter}:${range}`;
}

File: mobile/lib/services/bible.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config.dart';

class BookRef {
  final int num;
  final String name;
  BookRef(this.num, this.name);
}

class Chapter {
  final String chapter;
  final String title;
  final List<String> verses;
  Chapter(this.chapter, this.title, this.verses);
}

class Book {
  final String title;
  final List<Chapter> chapters;
  Book(this.title, this.chapters);
}

class BibleService {
  BibleService._();
  static final BibleService instance = BibleService._();

  final Map<int, Book> _cache = {};
  List<BookRef>? _books;

  Future<dynamic> _get(String path) async {
    Object? lastErr;
    for (var attempt = 0; attempt < 5; attempt++) {
      try {
        final res = await http
            .get(Uri.parse('${Config.bibleApi}$path'))
            .timeout(const Duration(seconds: 8));
        if (res.statusCode != 200) {
          throw Exception('Bible API ${res.statusCode}');
        }
        return jsonDecode(utf8.decode(res.bodyBytes));
      } catch (e) {
        lastErr = e;
        if (attempt < 4) {
          await Future.delayed(Duration(milliseconds: 300 * (attempt + 1)));
        }
      }
    }
    throw lastErr ?? Exception('Bible API failed for $path');
  }

  Future<List<BookRef>> getBooks() async {
    if (_books != null) return _books!;
    final arr = (await _get('/book') as List).cast<String>();
    final reg = RegExp(r'^(\d+)\s*:\s*(.+)$');
    final list = <BookRef>[];
    for (final s in arr.skip(1)) {
      final m = reg.firstMatch(s);
      if (m != null) {
        list.add(BookRef(int.parse(m.group(1)!), m.group(2)!.trim()));
      }
    }
    list.sort((a, b) => a.num - b.num);
    _books = list;
    return list;
  }

  Future<Book> getBook(int num) async {
    if (_cache.containsKey(num)) return _cache[num]!;
    final j = await _get('/book/$num') as Map<String, dynamic>;
    final chapters = (j['chapters'] as List)
        .map((c) => Chapter(
              (c['chapter'] ?? '').toString(),
              (c['title'] ?? '').toString(),
              (c['verses'] as List).map((v) => v.toString()).toList(),
            ))
        .toList();
    final book = Book((j['title'] ?? '').toString(), chapters);
    _cache[num] = book;
    return book;
  }

  String? bookNameSync(int num) {
    return _books?.where((b) => b.num == num).firstOrNull?.name;
  }
}

extension _FirstOrNull<E> on Iterable<E> {
  E? get firstOrNull => isEmpty ? null : first;
}
